twisted/txacme
# Initial support for ACME v2

# Scope

This adds support for ACME v2 and removes support for ACME v1.

It tries to fixes #151

The endpoints are not yet implemented for ACME v2... maybe remove them for the moment and add then back with the ALPN-01 validation.


# Changes

Some eliot structured logs were removed as I could not make sense of their structure and were just a distraction.,, maybe I will get a better understand of the eliot usage soon and the structure of the logs and will be able to update them.

txacme.service.AcmeIssuingService high-level API was not changed . ... but I don't like the way service starts without being able to catch account registration / update errors.

endpoint API was not changed.... but it is broken.
I would say that we can remove it and add it later... as otherwise this PR is too big.

At first, I want to have only txacme.ciient.Client low-level API in the scope of this PR...but then I found that txacme.service.AcmeIssuingService can be updated with little effort.

------------

txacme.ciient.Client low-level API is a complete rewrite as ACME v2 has different operations.

txacme.ciient.Client API which was not changed:

* txacme.ciient.Client.fromURL 
* txacme.ciient.Client.answer_challenge 

txacme.ciient.Client API which was removed:

* txacme.ciient.Client.retry_after
* txacme.ciient.Client.fetch_chain 


txacme.ciient.Client API which was replaced:

* txacme.ciient.Client.register and update_registration API was replaced by a `start` method... which works in tandem with stop()

* txacme.ciient.Client.request_challenges was replaced by `submit_order`.

* txacme.ciient.Client.request_issuance was replaced by `finalize` and fetch_certificate.

* txacme.ciient.Client.poll was replaced by `check_authorization` and `check_order` as you should now poll multiple resource types.

* txacme.ciient.poll_until_valid was removed and replaced with `get_certificate`


-----------

txacme.client.answer_challenge was updated to do a poll operation and match the get_certificate API

-----------

For now there is a custom Finalize request while https://github.com/certbot/certbot/pull/7825 is merged and released.


# To do

* [ ] Restore eliot structured logging
* [ ] Write automated unit tests
* [ ] Write some automated tests against pebble ACME v2 implementation... automate the tests from the example scripts


# How to test

I have tested it with pebble (including strict mode) and with the Let's Encrypt staging and prod servers.

1. Check that the examples make sense
2. Check that the news fragment are correct
3. Check that all the code is used and there is no code which is no longer used (maybe remove endpoint code)


## Contributor Checklist:

* [x] The Pull Request description explain the purpose of this PR or provides
  a link / reference to a GitHub Issue.
* [x] Created a newsfragment in src/txacme/newsfragments/.
* [ ] Updated the automated tests.
* [x] The changes pass minimal style checks.
